Firestore/core/src/firebase/firestore/remote/serializer.cc

```diff
@@ -509,13 +509,18 @@ std::map<std::string, FieldValue> DecodeObject(pb_istream_t* stream) {
 
 void Serializer::EncodeFieldValue(const FieldValue& field_value,
                                   std::vector<uint8_t>* out_bytes) {
-  // TODO(rsgowman): how large should the output buffer be? Do some
-  // investigation to see if we can get nanopb to tell us how much space it's
-  // going to need. (Hint: use a sizing stream, i.e. PB_OSTREAM_SIZING)
-  uint8_t buf[1024];
-  pb_ostream_t stream = pb_ostream_from_buffer(buf, sizeof(buf));
+  pb_ostream_t stream = {
+      /*callback=*/[](pb_ostream_t* stream, const pb_byte_t* buf,
+                      size_t count) -> bool {
+        auto* out_bytes = static_cast<std::vector<uint8_t>*>(stream->state);
+        out_bytes->insert(out_bytes->end(), buf, buf + count);
+        return true;
+      },
+      /*state=*/out_bytes,
+      /*max_size=*/SIZE_MAX,
+      /*bytes_written=*/0,
```

Comment:
Does this need to be `out_bytes->size()`? If not can you add comments on why these values work?

Reply:
The only thing that *really* matters is max_bytes - bytes_written, i.e. how many more bytes is nanopb allowed to write to this stream? So we could do either of these, and they'd be equivalent:

```
// (a)
max_bytes = 1MiB + out_bytes.size()
bytes_written = out_bytes.size()
// (b)
max_bytes = 1MiB
bytes_written = 0
```

I don't think I've ever seen bytes_written *not* initialized to 0. During operation, nanopb increments this as bytes are written. (If nanopb used a c++ object for this, I would expect the ctor to take a single parameter (max_bytes) and then always init bytes_written to 0.)

I've added some comments to help clarify.

Reply:
Hm. Well, not a *single* parameter. Lets rephrase to "I would expect to ctor to *not* take a bytes_written parameter"